fedimint-server/src/config/api.rs

```diff
@@ -1026,6 +1190,99 @@ mod tests {
+    async fn test_restart_setup() {
+        let _ = TracingSetup::default().init();
+        let (data_dir, _maybe_tmp_dir_guard) = test_dir("test-restart-setup");
+        let base_port = port_alloc(1).unwrap();
+
+        // let mut join_handles = vec![];
+        let mut apis = vec![];
+        let mut followers = vec![];
+        let (mut leader, api) = TestConfigApi::new(base_port, 0, data_dir.clone()).await;
+
+        apis.push(api);
+
+        for i in 1..4 {
+            let port = base_port + (i * 2);
+            let (follower, api) = TestConfigApi::new(port, i, data_dir.clone()).await;
+            apis.push(api);
+            followers.push(follower);
         }
+
+        // Run the Fedimint servers and test concurrently
+        spawn("Fedimint server apis", async move {
+            join_all(apis.iter_mut().map(|api| api.run(TaskGroup::new()))).await;
+        });
+
+        assert_eq!(leader.status().await.server, ServerStatus::AwaitingPassword);
+
+        leader
+            .client
+            .set_password(leader.auth.clone())
+            .await
+            .unwrap();
+
+        leader.set_connections(&None).await.unwrap();
+        leader.name = "leader".to_string();
+        leader.set_connections(&None).await.unwrap();
+
+        // Leader sets the config
+        let _ = leader
+            .client
+            .get_default_config_gen_params(leader.auth.clone())
+            .await
+            .unwrap();
+        leader.set_config_gen_params().await;
+
+        // Setup followers and send connection info
+        for follower in &mut followers {
+            assert_eq!(
+                follower.status().await.server,
+                ServerStatus::AwaitingPassword
+            );
+            follower
+                .client
+                .set_password(follower.auth.clone())
+                .await
+                .unwrap();
+            let leader_url = Some(leader.settings.api_url.clone());
+            follower.set_connections(&leader_url).await.unwrap();
+            follower.name = format!("{}_", follower.name);
+            follower.set_connections(&leader_url).await.unwrap();
+            follower.set_config_gen_params().await;
+        }
+        leader
+            .wait_status(ServerStatus::SharingConfigGenParams)
+            .await;
+
+        // Leader can trigger a setup restart
+        leader
+            .client
+            .restart_federation_setup(leader.auth.clone())
+            .await
+            .unwrap();
+
+        // All peers can trigger a setup restart. This has to be done manually by each
+        // peer, and any peer could trigger a restart before the leader does.
+        for peer in &followers {
+            peer.client
+                .restart_federation_setup(peer.auth.clone())
+                .await
+                .ok();
+        }
+
+        leader.wait_status(ServerStatus::SetupRestarted).await;
+        leader.wait_status(ServerStatus::AwaitingPassword).await;
```

Comment:
Can we test if a second try of the setup will actually run through? E.g. also test what happens if one peer had a misconfigured API/p2p address with which they registered with the leader and need to correct it.

Reply:
now rerunning setup steps after restart